launch_testing/examples/test_pub_sub_launch.py

```diff
@@ -0,0 +1,230 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Some example tests around ROS 2 talker/listener nodes.
+
+.. note::
+
+To maximize code reuse, this test assumes that execution and executable
+abstractions are decoupled (see https://github.com/ros2/launch/issues/114).
+"""
+
+import sys
+import pytest
+
+from launch import LaunchDescription
+from launch.actions import Execute
+from launch.events.process import ProcessStarted
+from launch.launch_description_sources import \
+    get_launch_description_from_python_launch_file
+from launch_ros import get_default_launch_description
+from launch_ros.executables import Node
+from launch_ros.executables import ComposableNode
+from launch_ros.executables import ComposableNodeProcess
+import launch_ros.events.lifecycle as lifecyle
+
+from launch_testing import TestLaunchService
+from launch_testing.actions import PyTest
+from launch_testing.actions import GTest
+from launch_testing.actions import Assert
+from launch_testing.actions import AssertSequenceOnce
+
+import launch_testing.predicates as predicates
+import launch_testing.variables as variables
+
+
+@pytest.fixture
+def ros_test_launch_service() -> TestLaunchService:
+    """
+    A test fixture providing a test-aware launch service (i.e. one
+    that listens to events fired by `launch_testing.actions.Assert`
+    and `launch_testing.actions.Test` actions and reacts accordingly).
+    """
+    ls = TestLaunchService()
+    # TODO(hidmic): consider integrating ROS specific actions and the
+    # launch service in some way to ease launch file composition without
+    # repetition).
+    ls.include_launch_description(
+        get_default_launch_description(prefix_output_with_name=False)
+    )
+    return ls
+
+
+@pytest.fixture
+def pub_sub_launch_description() -> LaunchDescription:
+    """
+    A test fixture providing the basic pub/sub launch description.
+    """
+    # TODO(hidmic): implement functionality to lookup launch files
+    # on a per package basis
+    launch_path = get_launch_path(
+        package_name='launch_ros', launch_file_path='pub_sub_launch.py'
+    )
+    return get_launch_description_from_python_launch_file(launch_path)
+
+
+def test_pub_sub_launch(ros_test_launch_service, pub_sub_launch_description):
+    """
+    Tests an existing simple pub/sub system.
+    """
+    # Launches a pytest that also happens to be a node with
+    # a 30s timeout.
+    pub_sub_launch_description.add_action(
+        # TODO(hidmic): allow launching non-installed
+        # executables within the current package.
+        PyTest(Node(
```

Comment:
I don't think `Node` is the write argument for `PyTest` (or `GTest` below). The actual Python script containing tests or the executable containing a gtest main function are not necessarily nodes.

Reply:
That's exactly the point I want to make. Both the pytest script and the gtest executable could be nodes, containers of composable nodes, plain ROS-agnostic programs, etc. So we'd want to reuse what we have in terms of doing file lookups, argument passing, output handling, etc. If you think about it, all the `PyTest` and `GTest` actions have to do is execute whatever is being passed in a single host process (for starters) assuming it's what we say it is, pass specific command line arguments and collect whatever output it's expected the executable will generate. 

So above, where it says `Node`, it could say `Process` (i.e. any non ROS aware executable).

Reply:
> So above, where it says `Node`, it could say `Process` (i.e. any non ROS aware executable).

Maybe I am missing how this is supposed to work.

A Python script being a node will have a `__main__` which get the node running. On the other hand a Python file containing unit tests invoked by `pytest` doesn't. Also such a test file wouldn't be represented by `Process` (since it doesn't have a `__main__`). In fact a Python test file can't even be invoked - it commonly doesn't have an executable flag. So e.g. `node_executable` doesn't apply here - neither can you pass command line arguments of any kind.

Reply:
Specifically for pytest, I think Dirk might be right. For gtest, I do think it makes sense to have an argument that inherits from the proposed `Executable` or `ExecuteProcess` (as `Node` does).

Reply:
>A Python script being a node will have a __main__ which get the node running. On the other hand a Python file containing unit tests invoked by pytest doesn't. Also such a test file wouldn't be represented by Process (since it doesn't have a __main__).

That's mostly true, though not strictly. The following script can be made executable to run all the tests in it:

```python
#! /usr/bin/env python                                                          
                                                                                
import os                                                                       
import pytest                                                                   
                                                                                
                                                                                
def test_stuff():                                                               
    assert False                                                                
                                                                                
                                                                                
pytest.main([os.path.realpath(__file__)])
```

But, set aside above's unusual setup, you're right about `pytest` test files not being executable in general. And we do have a problem with a command line arguments, though [not insurmountable](https://docs.pytest.org/en/latest/reference.html#_pytest.hookspec.pytest_addoption).

--------

The initial idea was to use the fact that the testing frameworks 

So, as I see it, this leaves us with either of two options (that I'm going to sketch for both `gtest` and `pytest` as I'd rather have a consistent API):

1. We derive a _test_ version of each executable kind we have and we think of testing frameworks as testing protocols (e.g. how to execute, how to collect results, etc.):

  ```python
  NodeTest(
       path='path/to/test.py', protocol='pytest', 
       timeout=20, remapping=[('chatter', 'my_chatter')]
   ),
  ComposableNodeTest(
       container_executable='intra_proc_comm_test_container',
       composable_node_descriptions=[
           ComposableNode(
               package_name='composition',
               node_plugin_name='talker_component',
               name='my_talker_component'
           ),
           ComposableNode(
               package_name='composition',
               node_plugin_name='listener_component'
               name='my_listener_component'
           ),
        protocol='gtest', timeout=10
     ]
  )
  ```
  
   
2. We outright ban test parameterization at the launch level.

  ```python
  PyTest(path='path/to/test.py', timeout=10),
  GTest(path='path/to/gtest', timeout=10)
  ```

(1) is somewhat complex, but it enables full launch API reuse. (2) is easy to use and to implement, though a bit limiting in my opinion.

I guess the question is whether we want to provide all the tools to parameterize tests or if we just to want to cover the basic use cases and then leave the users on their own to come up with custom ways of doing so.

Reply:
At the moment I am not aware of any use case where we need `ComposableNodeTest`. In the future there could certainly be test using that approach. For now I would rather focus on the cases which we do have a use case for.

Reply:
Alright, went with the simple one.